File: algorithms/python3/0783-minimum-distance-between-bst-nodes.py
# 783. Minimum Distance Between BST Nodes
# https://leetcode.com/problems/minimum-distance-between-bst-nodes

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def minDiffInBST(self, root: Optional[TreeNode]) -> int:
    self.min = float('inf')
    self.prev = None
    self.inorder(root)
    return self.min

  def inorder(self, root: TreeNode) -> None:
    if root is None:
      return

    self.inorder(root.left)
    if self.prev is not None:
      self.min = min(self.min, root.val - self.prev)
    self.prev = root.val
    self.inorder(root.right)

File: algorithms/ruby/0783-minimum-distance-between-bst-nodes.rb
# frozen_string_literal: true

# 783. Minimum Distance Between BST Nodes
# https://leetcode.com/problems/minimum-distance-between-bst-nodes
# Easy
# TODO: Write Tests

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def min_diff_in_bst(root)
  @prev = nil
  @min = nil
  inorder_traversal(root)
  @min
end

def inorder_traversal(root)
  return if root.nil?
  inorder_traversal(root.left)
  if @prev
    val = root.val - @prev
    @min = @min ? [@min, val].min : val
  end
  @prev = root.val
  inorder_traversal(root.right)
end
